docs/rfc/rfc-011-delete-gas.md

```diff
@@ -0,0 +1,140 @@
+parameter.
+
+Gas has gained adoption within the Cosmos ecosystem [as part of the Cosmos SDK][cosmos-sdk-gas].
+The SDK provides facilities for tracking how much 'Gas' a transaction is expected to take
+and a mechanism for tracking how much gas a transaction has already taken.
+
+Non-SDK applications also make use of the concept of Gas. Anoma appears to implement
+[a gas system][anoma-gas] to meter the transactions it executes.
+
+While the notion of gas is present in projects that make use of Tendermint, it is
+not a concern of Tendermint's. Tendermint's value and goal is producing blocks
+via a distributed consensus algorithm. Tendermint relies on the application specific
+code to decide how to handle the transactions Tendermint has produced (or if the
+application wants to consider them at all). Gas is an application concern.
+
+Our implementation of Gas is causing issues for node operators and relayers. At
+the moment, transactions that overflow the configured 'MaxGas' can be silently rejected
+from the mempool. Overflowing MaxGas is the _only_ way that a transaction can be considered
+invalid that is not directly a result of failing the `CheckTx`. Operators, and the application,
+do not know that a transaction was removed from the mempool for this reason. A stateless check
+of this nature is exactly what `CheckTx` exists for and there is no reason for the mempool
+to keep track of this data separately.
+
+Removing Gas from Tendermint and the mempool would allow for the 'mempool' to be a better
+abstraction: any transaction that arrived at `CheckTx` and passed the check will either be
+a candidate for a later block or evicted after a TTL is reached. All other transactions are 
+completely invalid and can be discarded forever.
+
+Removing gas will not be completely straightforward. It will mean ensuring that 
+equivalent functionality can be implemented outside of the mempool using the mempool's API.
+
+## Discussion
+
+This section catalogs the functionality that will need to exist within the Tendermint
+mempool to allow Gas to be removed and replaced by application-side bookkeeping.
+
+### Requirement: Provide Mempool Tx Sorting Mechanism
+
+Gas produces a market for inclusion in a block. On many networks, a [gas fee][cosmos-sdk-fees] is
+included in pending transactions. This fee indicates how much a user is willing to
+pay per unit of execution and the fees are distributed to validators.
+
+Validators wishing to extract higher gas fees are incentivized to include transactions 
+with the highest listed gas fees into each block. This produces a naturally ordering
+of the pending transactions. Applications wishing to implement a gas mechanism need
+to be able to order the transactions in the mempool. This can trivially be accomplished 
+by sorting transactions using the `priority` field available to applications as part of
+v0.35's `ResponseCheckTx` message.
```

Comment:
I think it's tempting to throw up our hands at the knapsack porblem and be like "this is just NP-hard (complete?) so why bother," when I think it's possible to do pretty well.

Reply:
> which transactions to pick (and how many, and how big, etc.) is fundamentally not that interesting for Tendermint to care about, and the application should probably decide it.

At first my reaction was strong disagreement with this, but I think I'm coming around to it. TM's job is to ensure that the network agrees that a set of Tx is fitting to execute next. What constitutes a 'full' block is likely not its concern. Full is really related to how many Tx an application thinks it can execute and how quickly the network can gossip them. 

Different applications may wish to implement more interesting or advanced strategies to determine this: for example implementing some setpoint of block or tx throughput.

Reply:
I think Michael has some very convincing arguments. 

A lot of this seems to have to do with message framing or in this case block framing. We need to design a system that buffers an uneven load of transactions on one side and frames them into blocks on the other. Most of the times we frame based on time. If we have too many transactions then we frame by bytes. Gas is the third factor we use for framing which corresponds to computation, networking bandwidth and storage.

Timing is a bit difficult for an app to control because it comes down to the rhythm of consensus. As to the other two, Prior to ABCI++, the only way the application could control them was through `ConsensusParams` in Tendermint. So Tendermint had to know about them. With ABCI++, control of framing blocks can be directly passed on to the application itself. We could decide now that both `MaxBytes` and `MaxGas` are constraints that Tendermint doesn't need to keep track of. 

There is an argument however to counter that. Tendermint directly controls the p2p layer and thus it's responsible for mitigating DoS attacks. Knowing the size of a block is something Tendermint can do without knowing anything inside the transactions. This can prevent a malicious node from sending excessively large blocks (It's possible that consensus timeouts may also prevent this but I'm not sure). So perhaps we still need some upper bound in place.

Whether a block has too much Gas is something only the application knows. I currently don't know how Tendermint  prevents consensus from agreeing on a block with too much Gas.

Reply:
> Whether a block has too much Gas is something only the application knows. I currently don't know how Tendermint prevents consensus from agreeing on a block with too much Gas.

It appears from the discussions in slack that it _doesn't_ and in fact that, given the current constraint of ABCI, it cannot reject a block on this ground. I would say that this is arguably an even stronger reason to remove Gas from the mempool. It is a constraint we have placed on blocks via their creation, but not via their propagation and it is a constraint that TM can never enforce without input from the app on every Tx.

Reply:
Added info to this effect to the RFC.